src/tests.rs

```diff
@@ -0,0 +1,88 @@
+use crate::{self as pallet_template, Config };
+use frame_support::{assert_noop, assert_ok, construct_runtime, parameter_types,
+                    traits::Everything, traits::Currency};
+use sp_core::H256;
+use sp_io::TestExternalities;
+use sp_runtime::{
+    testing::Header,
+    traits::{BlakeTwo256, IdentityLookup},
+};
+
+type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<TestRuntime>;
+type Block = frame_system::mocking::MockBlock<TestRuntime>;
+
+construct_runtime!(
+	pub enum TestRuntime where
+		Block = Block,
+		NodeBlock = Block,
+		UncheckedExtrinsic = UncheckedExtrinsic,
+	{
+		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
+		PalletTemplate: pallet_template::{Pallet, Call, Storage, Event<T>},
+	}
+);
+
+parameter_types! {
+	pub const BlockHashCount: u64 = 250;
+	pub BlockWeights: frame_system::limits::BlockWeights =
+		frame_system::limits::BlockWeights::simple_max(1024);
+}
+impl frame_system::Config for TestRuntime {
+    type BaseCallFilter = Everything;
+    type BlockWeights = ();
+    type BlockLength = ();
+    type Origin = Origin;
+    type Call = Call;
+    type Index = u64;
+    type BlockNumber = u64;
+    type Hash = H256;
+    type Hashing = BlakeTwo256;
+    type AccountId = u64;
+    type Lookup = IdentityLookup<Self::AccountId>;
+    type Header = Header;
+    type Event = Event;
+    type BlockHashCount = BlockHashCount;
+    type DbWeight = ();
+    type Version = ();
+    type PalletInfo = PalletInfo;
+    type AccountData = ();
+    type OnNewAccount = ();
+    type OnKilledAccount = ();
+    type SystemWeightInfo = ();
+    type SS58Prefix = ();
+    type OnSetCode = ();
+    type MaxConsumers = frame_support::traits::ConstU32<16>;
+}
+
+impl Config for TestRuntime {
+    type Event = Event;
+    type Currency = ();
+    type MaxMetadataLength = ();
+}
+
+struct ExternalityBuilder;
+
+impl ExternalityBuilder {
+    pub fn build() -> TestExternalities {
+        let storage = frame_system::GenesisConfig::default()
+            .build_storage::<TestRuntime>()
+            .unwrap();
+        let mut ext = TestExternalities::from(storage);
+        ext.execute_with(|| System::set_block_number(1));
+        ext
+    }
+}
+
+
+#[test]
+fn subscribe(){
+    ExternalityBuilder::build().execute_with(|| {
+        assert_ok!(PalletTemplate::subscribe(Origin::signed(1),2,4000,5));
+
+        assert_noop!(PalletTemplate::subscribe(Origin::signed(1),2,0,5), "frequency or amount is 0");
+
+        let expected_event = Event::PalletTemplate(pallet_template::Event::SubscriptionStored (5, 4000));
+
+        assert_eq!(System::events()[0].event, expected_event);
+    })
+}
```

Comment:
Do more tests. Write a test for each case where it can fail
Write tests for successive calls